pyuvdata/uvbeam/initializers.py

```diff
@@ -0,0 +1,380 @@
+            )
+
+        if element_location_array.ndim != 2:
+            raise ValueError("element_location_array must be 2 dimensional")
+        shape = element_location_array.shape
+        if shape[0] != 2:
+            raise ValueError(
+                "The first dimension of element_location_array must be length 2"
+            )
+        if shape[1] <= 1:
+            raise ValueError(
+                "The second dimension of element_location_array must be >= 2."
+            )
+
+        if element_coordinate_system is None:
+            element_coordinate_system = "x-y"
+        if delay_array is None:
+            delay_array = np.zeros(shape[1])
+        else:
+            if delay_array.shape != (shape[1],):
+                raise ValueError(
+                    "delay_array must be one dimensional with length "
+                    "equal to the second dimension of element_location_array"
+                )
+        if gain_array is None:
+            gain_array = np.ones(shape[1])
+        else:
+            if gain_array.shape != (shape[1],):
+                raise ValueError(
+                    "gain_array must be one dimensional with length "
+                    "equal to the second dimension of element_location_array"
+                )
+        coupling_shape = (shape[1], shape[1], uvb.Nfeeds, uvb.Nfeeds, uvb.Nfreqs)
+        if coupling_matrix is None:
+            coupling_matrix = np.zeros(coupling_shape, dtype=complex)
+            for element in range(shape[1]):
+                coupling_matrix[element, element] = np.ones(
+                    (uvb.Nfeeds, uvb.Nfeeds, uvb.Nfreqs), dtype=complex
+                )
+        else:
+            if coupling_matrix.shape != coupling_shape:
+                raise ValueError(
+                    f"coupling_matrix shape {coupling_matrix.shape} does not "
+                    f"match expected shape {coupling_shape}."
+                )
+
+        uvb.antenna_type = "phased_array"
+        uvb._set_phased_array()
+        uvb.element_coordinate_system = element_coordinate_system
+        uvb.element_location_array = element_location_array
+        uvb.Nelements = shape[1]
+        uvb.delay_array = delay_array
+        uvb.gain_array = gain_array
+        uvb.coupling_matrix = coupling_matrix
+    else:
+        uvb.antenna_type = "simple"
+        uvb._set_simple()
+
+    # Set data parameters
+    if uvb.beam_type == "efield":
+        if uvb.pixel_coordinate_system == "healpix":
+            data_shape = (uvb.Naxes_vec, uvb.Nfeeds, uvb.Nfreqs, uvb.Npixels)
+        else:
+            data_shape = (uvb.Naxes_vec, uvb.Nfeeds, uvb.Nfreqs, uvb.Naxes2, uvb.Naxes1)
+        data_type = complex
+    else:
+        if uvb.pixel_coordinate_system == "healpix":
+            data_shape = (uvb.Naxes_vec, uvb.Npols, uvb.Nfreqs, uvb.Npixels)
+        else:
+            data_shape = (uvb.Naxes_vec, uvb.Npols, uvb.Nfreqs, uvb.Naxes2, uvb.Naxes1)
+        data_type = float
```

Comment:
This could be tighter:

```python
if uvb.beam_type == "efield":
    data_type = complex
    polax = uvb.Nfeeds
else: 
    data_type = float
    polax = uvb.Npols

if uvb.pixel_coordinate_system == "healpix":
    pixax = (uvb.Npixels,)
else:
    pixax = (uvb.Naxes2, uvb.Naxes1)

shape = (uvb.Naxes_vec, polax, uvb.Nfreqs) + pixax
```